### PR TITLE
[SPARK-20763][SQL]The function of `month` and `day`  return the value which is not we expected.

### PR DESCRIPTION
**Pull request by 10110346**
## What changes were proposed in this pull request?
spark-sql>select month("1582-09-28");
spark-sql>10
For this case, the expected result is 9, but it is 10.

spark-sql>select day("1582-04-18");
spark-sql>28
For this case, the expected result is 18, but it is 28.

when the date  before "1582-10-04", the function of `month` and `day` return the value which is not we expected.

## How was this patch tested?
unit tests

### CODE REVIEW

**Comment by srowen** [comment]
Point taken, but do other DBs support dates before 1970? it does highly depend on historical calendars. The right solution isn't a hack in one place here but using Calendar properly, if anything.

**Comment by 10110346** [comment]
@srowen I have tested in mysql, it can support dates before 1970.
mysql> select month("1582-09-28");
+---------------------+
| month("1582-09-28") |
+---------------------+
|                   9              |
+---------------------+

mysql> select day("1582-04-18");
+-------------------+
| day("1582-04-18") |
+-------------------+
|                18          |
+-------------------+

**Comment by 10110346** [comment]
I have tried to changed `getYearAndDayInYear` like this:
`private[this] def getYearAndDayInYear(daysSince1970: SQLDate): (Int, Int, Int) = {
    val date = new Date(daysToMillis(daysSince1970))
    val YMD = date.toString.split("-")
    (YMD(0).toInt, YMD(1).toInt, YMD(2).toInt)
  }`
The result is right, but 'checkConsistencyBetweenInterpretedAndCodegen' executed failed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 1]
```diff
@@ -601,22 +601,32 @@ object DateTimeUtils {
...
-    // add the difference (in days) between 1.1.1970 and the artificial year 0 (-17999)
-    val daysNormalized = daysSince1970 + toYearZero
-    val numOfQuarterCenturies = daysNormalized / daysIn400Years
-    val daysInThis400 = daysNormalized % daysIn400Years + 1
-    val (years, dayInYear) = numYears(daysInThis400)
-    val year: Int = (2001 - 20000) + 400 * numOfQuarterCenturies + years
-    (year, dayInYear)
+  private[this] def getYearAndDayInYear(daysSince1970: SQLDate): (Int, Int, Int) = {
+    val date = new Date(daysToMillis(daysSince1970))
+    val YMD = date.toString.trim.split("-")
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: this would cause massive performance regression.

Comment (10110346) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: yes,this cause performance worse than before

**Comment by 10110346** [comment]
@srowen @rxin  So, by contrast, mybe  make a hack in one place is a better solution

**Comment by ueshin** [comment]
I guess this would also affect date-related functions such as `dayofyear`, `year`, `quater` and `weekofyear`, etc.
Could you add tests for them?

**Comment by 10110346** [comment]
@ueshin  Yes, I will do it ,thanks

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 2]
```diff
@@ -76,6 +76,9 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       }
     }
     checkEvaluation(DayOfYear(Literal.create(null, DateType)), null)
+
+    checkEvaluation(DayOfYear(Literal(new Date(sdf.parse("1582-04-28 13:10:15").getTime))), 118)
+    checkEvaluation(DayOfYear(Literal(new Date(sdf.parse("1582-10-4 13:10:15").getTime))), 277)
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: I believe the dates we need to test here should be the boundaries like `1582-10-04` and `1582-10-15`.

Btw, let's add `0` for 1-digit date, e.g. `1582-10-04` instead of `1582-10-4`.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 3]
```diff
@@ -96,6 +99,8 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
         }
       }
     }
+    checkEvaluation(Year(Literal(new Date(sdf.parse("1582-01-01 13:10:15").getTime))), 1582)
+    checkEvaluation(Year(Literal(new Date(sdf.parse("1582-12-31 13:10:15").getTime))), 1582)
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: Here should be `1581-12-31` and `1582-01-01`, for example, and the rest should be in a similar way.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 4]
```diff
@@ -116,6 +121,9 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
         }
       }
     }
+
+    checkEvaluation(Quarter(Literal(new Date(sdf.parse("1582-03-28 13:10:15").getTime))), 1)
+    checkEvaluation(Quarter(Literal(new Date(sdf.parse("1582-09-30 13:10:15").getTime))), 3)
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: How about using `1582-09-30` and `1582-10-01` for this?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala [hunk 5]
```diff
@@ -76,6 +76,9 @@ class DateExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       }
     }
     checkEvaluation(DayOfYear(Literal.create(null, DateType)), null)
+
+    checkEvaluation(DayOfYear(Literal(new Date(sdf.parse("1582-10-15 13:10:15").getTime))), 288)
```
Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: I'm not sure whether this value is correct or not. should be `278`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 6]
```diff
@@ -603,7 +603,13 @@ object DateTimeUtils {
    */
   private[this] def getYearAndDayInYear(daysSince1970: SQLDate): (Int, Int) = {
     // add the difference (in days) between 1.1.1970 and the artificial year 0 (-17999)
-    val daysNormalized = daysSince1970 + toYearZero
+    var  daysSince1970Tmp = daysSince1970
+    // In history,the period(5.10.1582 ~ 14.10.1582) is not exist
```
Comment (kiszk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: nit: this representation may be confusing for all of the people in the world. Can we use 1582-10-5 or like Oct. 5, 1582?

Comment (10110346) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: I have tested like this:
`val dav = Date.valueOf("1582-10-14")
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is:`1582-10-24`
`val dav = Date.valueOf("1582-10-05")
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is:`1582-10-15`
so, i think we can not use `1582-10-5` ~ `1582-10-14`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 5 above)
Comment (10110346) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: I'm not sure too, maybe `278` is better

Comment (ueshin) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: cc @cloud-fan @gatorsmile
Do you have any ideas?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: It's only about comment, and I think 1582-10-5 or Oct. 5, 1582 is more human readable.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala (hunk 5 above)
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: can we check with other databases?

Comment (10110346) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: In mysql ,the rusult is :
mysql> select dayofyear("1982-10-04");
+-------------------------+
| dayofyear("1982-10-04") |
+-------------------------+
|                     277 |
+-------------------------+
1 row in set (0.00 sec)

mysql> select dayofyear("1982-10-015");
+--------------------------+
| dayofyear("1982-10-015") |
+--------------------------+
|                      288 |
+--------------------------+

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: why `278` is better?

Comment (10110346) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: @cloud-fan Because in history,the period(5.10.1582 ~ 14.10.1582) is not exist

Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: let's follow mysql

Comment (10110346) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/DateExpressionsSuite.scala]: OK, thanks @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (10110346) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: OK, I will do ,thanks @kiszk @cloud-fan

**Comment by ueshin** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: `In history,the period(5.10.1582 ~ 14.10.1582) is not exist`
->
`Since Julian calendar was replaced with the Gregorian calendar, the 10 days after Oct. 4 were skipped.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 7]
```diff
@@ -603,7 +603,13 @@ object DateTimeUtils {
    */
   private[this] def getYearAndDayInYear(daysSince1970: SQLDate): (Int, Int) = {
     // add the difference (in days) between 1.1.1970 and the artificial year 0 (-17999)
-    val daysNormalized = daysSince1970 + toYearZero
+    var  daysSince1970Tmp = daysSince1970
+    // In history,the period(1582-10-05 ~ 1582-10-14) is not exist
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: `In history,the period(5.10.1582 ~ 14.10.1582) is not exist`
->
`Since Julian calendar was replaced with the Gregorian calendar, the 10 days after Oct. 4 were skipped.`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 6 above)
Comment (10110346) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @gatorsmile ok,thanks

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala (hunk 7 above)
Comment (10110346) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: @gatorsmile ok,thanks

**Comment by gatorsmile** [comment]
The current impl is still missing the special handling of that miracle period. See the outputs of Oracle:
```
Result Set 34
TO_CHAR(TO_DATE('1582-10-04','YYYY-MM-DD'),'DDD')
277

Result Set 35
TO_CHAR(TO_DATE('1582-10-05','YYYY-MM-DD'),'DDD')
288

Result Set 36
TO_CHAR(TO_DATE('1582-10-11','YYYY-MM-DD'),'DDD')
288

Result Set 37
TO_CHAR(TO_DATE('1582-10-12','YYYY-MM-DD'),'DDD')
288

Result Set 38
TO_CHAR(TO_DATE('1582-10-13','YYYY-MM-DD'),'DDD')
288

Result Set 39
TO_CHAR(TO_DATE('1582-10-14','YYYY-MM-DD'),'DDD')
288

Result Set 40
TO_CHAR(TO_DATE('1582-10-15','YYYY-MM-DD'),'DDD')
288

Result Set 41
TO_CHAR(TO_DATE('1582-10-16','YYYY-MM-DD'),'DDD')
289
```

**Comment by 10110346** [comment]
@gatorsmile  I have tested like this:
  `val dav = Date.valueOf("1582-10-04");
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is :1582-10-04
  `val dav = Date.valueOf("1582-10-05");
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is :1582-10-15
`val dav = Date.valueOf("1582-10-14");
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is :1582-10-24
`val dav = Date.valueOf("1582-10-15");
  val date = new Date(dav.getTime);
  println(date.toString)`
the output is :1582-10-15

Also,i tested in mysql:
`mysql> select dayofyear('1582-10-4');
+------------------------+
| dayofyear('1582-10-4') |
+------------------------+
|                    277 |
+------------------------+
1 row in set (0.00 sec)

mysql> select dayofyear('1582-10-5');
+------------------------+
| dayofyear('1582-10-5') |
+------------------------+
|                    278 |
+------------------------+
1 row in set (0.00 sec)

mysql> select dayofyear('1582-10-6');
+------------------------+
| dayofyear('1582-10-6') |
+------------------------+
|                    279 |
+------------------------+
1 row in set (0.00 sec)

mysql> select dayofyear('1582-10-14');
+-------------------------+
| dayofyear('1582-10-14') |
+-------------------------+
|                     287 |
+-------------------------+
1 row in set (0.00 sec)

mysql> select dayofyear('1582-10-15');
+-------------------------+
| dayofyear('1582-10-15') |
+-------------------------+`

**Comment by gatorsmile** [comment]
After checking DB2, it behaves the same as MySQL. 

LGTM

**Comment by gatorsmile** [comment]
Thanks! Merging to master/2.2.

**Comment by gatorsmile** [comment]
Could you submit a backport PR to 2.1? Thanks!

**Comment by 10110346** [comment]
OK，l Will do it, thanks. @gatorsmile
